File: Disjoint Set/20040번 - 사이클 게임/20040.py
import sys
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

n, m = map(int, input().split())
parents = list(range(n))
    
def Find(x):
    if x != parents[x]: 
        parents[x] = Find(parents[x])
    return parents[x]
    
def Union(a, b):
    a = Find(a)
    b = Find(b)
    if a != b:
        parents[b] = a
    
def solve():
    for i in range(1, m + 1):
        a, b = map(int, input().split())
        if Find(a) == Find(b):
            print(i)
            return
        Union(a, b)
    print(0)

if __name__ == '__main__':
    solve()

File: Disjoint Set/20040번 - 사이클 게임/20040.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, m;
vector<pair<int, int> > line;
int parents[500001];
int vis[500001];

void input() {
    cin >> n >> m;
    for(int i = 0, a, b; i < m; i++) {
        cin >> a >> b;
        line.push_back({a, b});
    }
    for(int i = 0; i < n; i++) {
        parents[i] = i;
    }
}

int Find(int x) {
    if(x == parents[x]) return x;
    return parents[x] = Find(parents[x]);
}

void Union(int a, int b) {
    a = Find(a);
    b = Find(b);
    if(a != b) parents[b] = a;
}

void solve() {
    for(int i = 0; i < m; i++) {
        if(Find(line[i].first) == Find(line[i].second)) {
            cout << i + 1 << '\n';
            return;
        }

        Union(line[i].first, line[i].second);
    }
    
    cout << 0 << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}